internal/source/source.go

```diff
@@ -0,0 +1,184 @@
+package source
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/kubeshop/botkube/internal/plugin"
+	"github.com/kubeshop/botkube/pkg/bot/interactive"
+	"github.com/kubeshop/botkube/pkg/config"
+	"github.com/kubeshop/botkube/pkg/notifier"
+)
+
+// Dispatcher watches for enabled sources events and send them to notifiers.
+type Dispatcher struct {
+	log       logrus.FieldLogger
+	notifiers []notifier.Notifier
+	manager   *plugin.Manager
+	cfg       *config.Config
+	starter   func(ctx context.Context, pluginName string, pluginConfigs []any, sources []string) error
+}
+
+// NewDispatcher create a new Dispatcher instance.
+func NewDispatcher(log logrus.FieldLogger, notifiers []notifier.Notifier, manager *plugin.Manager, cfg *config.Config) *Dispatcher {
+	d := &Dispatcher{
+		log:       log,
+		notifiers: notifiers,
+		manager:   manager,
+		cfg:       cfg,
+	}
+
+	// TODO: it allows us to mock it for unit-test
+	// will be removed once we will update the gRPC contract on source
+	// and passing list of configuration.
+	d.starter = d.start
+
+	return d
+}
+
+// Start starts all sources and dispatch received events.
+func (d *Dispatcher) Start(ctx context.Context) error {
+	// startProcesses holds information about started unique plugin processes
+	// We start a new plugin process each time we see a new order of source bindings.
+	// We do that because we pass the array of configs to each `Stream` method and
+	// the merging strategy for configs can depend on the order.
+	// As a result our key is e.g. ['source-name1;source-name2']
+	startProcesses := map[string]struct{}{}
+
+	startPlugin := func(bindSources []string) error {
```

Comment:
Isn't this function complex enough to be named / defined separately?

Reply:
yes, you are right. I changed it to become a dedicated method on this struct 👍 

In the next PR, I will change it even more to make it more straight forward when it comes to the responsibilities. Because I see that it is to complex to maintain and test all those pieces together.